webapp/channels/src/components/admin_console/access_control/editors/cel_editor/editor.tsx

```diff
@@ -0,0 +1,363 @@
+    overviewRulerBorder: false,
+    quickSuggestions: false,
+    renderControlCharacters: false,
+    scrollbar: {
+        horizontal: 'hidden',
+        useShadows: false,
+    },
+    selectionHighlight: false,
+    showFoldingControls: 'never',
+    suggestOnTriggerCharacters: true,
+    unicodeHighlight: {
+        ambiguousCharacters: false,
+        invisibleCharacters: false,
+    },
+    unusualLineTerminators: 'auto',
+    wordWrapColumn: 400,
+    wrappingIndent: 'none',
+    wrappingStrategy: 'advanced',
+    contextmenu: false,
+};
+
+interface CELEditorProps {
+    value: string;
+    onChange: (value: string) => void;
+    onValidate?: (isValid: boolean) => void;
+    placeholder?: string;
+    className?: string;
+    userAttributes: Array<{
+        attribute: string;
+        values: string[];
+    }>;
+}
+
+// TODO: this is just a sample schema for the editor, we need to get the actual schema from the server
+
+function CELEditor({
+    value,
+    onChange,
+    onValidate,
+    placeholder = 'user.attributes.<attribute> == <value>',
+    className = '',
+    userAttributes,
+}: CELEditorProps): JSX.Element {
+    const [editorState, setEditorState] = useState({
+        expression: value,
+        isValidating: false,
+        isValid: true,
+        cursorPosition: {line: 1, column: 1},
+        validationErrors: [] as string[],
+        statusBarColor: 'var(--button-bg)',
+        showTestResults: false,
+        testResults: null as AccessControlTestResult | null,
+    });
+
+    const schemas = {
+        user: ['attributes'],
+        'user.attributes': userAttributes.map((attr) => attr.attribute),
+    };
+
+    const editorRef = useRef(null);
+    const monacoRef = useRef<monaco.editor.IStandaloneCodeEditor | null>(null);
+    const [showHelpModal, setShowHelpModal] = useState(false);
+
+    useEffect(() => {
+        setEditorState((prev) => ({...prev, expression: value}));
+    }, [value]);
+
+    useEffect(() => {
+        if (monacoRef.current && monacoRef.current.getValue() !== editorState.expression) {
+            monacoRef.current.setValue(editorState.expression);
+        }
+    }, [editorState.expression]);
+
+    const handleChange = useCallback((newValue: string) => {
+        setEditorState((prev) => ({
+            ...prev,
+            expression: newValue,
+            statusBarColor: 'var(--button-bg)',
+            validationErrors: [],
+        }));
+        onChange(newValue);
+    }, [onChange]);
+
+    const validateSyntax = useCallback(async () => {
+        setEditorState((prev) => ({...prev, isValidating: true}));
+
+        try {
+            const errors = await Client4.checkAccessControlExpression(editorState.expression);
+            const isValid = errors.length === 0;
+            setEditorState((prev) => ({
+                ...prev,
+                isValid,
+                validationErrors: errors.map((error) => `${error.message} @L${error.line}:${error.column + 1}`),
```

Comment:
(Post-merge) Does the server translate its error messages? I don't know how much control we have over them or if they could possibly be translated on the client to ensure they're in the user's language